plugins/modules/cloud/misc/proxmox_nic.py

```diff
@@ -0,0 +1,348 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright: (c) 2021, Lammert Hellinga (@Kogelvis) <lammert@hellinga.it>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+---
+module: proxmox_nic
+short_description: Management of NIC's for Qemu(KVM) VM's in a Proxmox VE cluster.
+version_added: 3.1.0
+description:
+  - Allows you to create/update/delete NIC's on Qemu(KVM) Virtual Machines in a Proxmox VE cluster.
+author: "Lammert Hellinga (@Kogelvis) <lammert@hellinga.it>"
+options:
+  bridge:
+    description:
+      - Add this interface to the specified bridge device. The Proxmox VE standard bridge is called C(vmbr0).
+    type: str
+  firewall:
+    description:
+      - Whether this interface should be protected by the firewall.
+    type: bool
+    default: false
+  interface:
+    description:
+      - Name of the interface, should be C(net[n]) where C(1 ≤ n ≤ 31).
+    type: str
+    required: true
+  link_down:
+    description:
+      - Whether this interface should be disconnected (like pulling the plug).
+    type: bool
+    default: false
+  mac:
+    description:
+      - C(XX:XX:XX:XX:XX:XX) should be a unique MAC address. This is automatically generated if not specified.
+      - When not specified this module will keep the MAC address the same when changing an existing interface.
+    type: str
+  model:
+    description:
+      - The NIC emulator model.
+    type: str
+    choices: ['e1000', 'e1000-82540em', 'e1000-82544gc', 'e1000-82545em', 'i82551', 'i82557b', 'i82559er', 'ne2k_isa', 'ne2k_pci', 'pcnet',
+              'rtl8139', 'virtio', 'vmxnet3']
+    default: virtio
+  mtu:
+    description:
+      - Force MTU, for C(virtio) model only. Set to C(1) to use the bridge MTU.
+      - Value should be C(1 ≤ n ≤ 65520).
+    type: int
+  name:
+    description:
+      - Specifies the VM name. Only used on the configuration web interface.
+      - Required only for I(state=present).
+    type: str
+  queues:
+    description:
+      - Number of packet queues to be used on the device.
+      - Value should be C(0 ≤ n ≤ 16).
+    type: int
+  rate:
+    description:
+      - Rate limit in MBps (MegaBytes per second) as floating point number.
+    type: float
+  state:
+    description:
+      - Indicates desired state of the NIC.
+    type: str
+    choices: ['present', 'absent']
+    default: present
+  tag:
+    description:
+      - VLAN tag to apply to packets on this interface.
+      - Value should be C(1 ≤ n ≤ 4094).
+    type: int
+  trunks:
```

Comment:
I couldn't find this option in https://pve.proxmox.com/pve-docs/api-viewer/index.html#/nodes/{node}/qemu/{vmid}/config.
Also IIRC proxmox can not have a limit on vlan tags allowed on an interface, you can just set a network of node in `VLAN Aware` mode, but you can not limit the vms using that bridge what vlan tags they can pass or not.

Reply:
The trunks options is mentioned on the page you linked under `net[n]` for PUT and POST requests.
It is also mentioned here: https://pve.proxmox.com/wiki/Manual:_qm.conf although the description is very brief.

In the proxmox PVE code the trunks option is parsed, this can be found in `/usr/share/perl5/PVE/Network.pm`
For bridge interfaces this happens around line 223
For OVS setups around line 249
(Line numbers on a Proxmox VE 6.3 system)

For normal bridge interfaces it will run `/sbin/bridge vlan add dev <iface> vid <trunk>` for every vlan id specified in trunks. Those will be added on the corresponding tap device on the host `tap<vmid>i<interface id>`.

From what I've read on several proxmox forum posts, if you want a guest to communicate to certain vlans you also have to set the trunks config. Those answers were from proxmox staff members.
My network setups are a bit different so I can't easily test if a VM would be able to communicate without this setting on a certain vlan. But I did see the trunks settings reflected in the interface and on the command-line using 

```
root@pm2 ~ # bridge vlan show dev tap103i0
port	vlan ids
tap103i0	1
	4 PVID Egress Untagged
	4094
```
which has this as (relevant) ansible playbook config:

```yaml
interface: "net0"
bridge: "vmbr1"
tag: 4
model: "e1000"
trunks:
  - 1
  - 4094
```

Reply:
Right, I didn't find and I didn't know of this option. Thanks for the detailed answer!

Reply:
You're most welcome :slightly_smiling_face: